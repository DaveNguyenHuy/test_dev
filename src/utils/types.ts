export interface Movie {
  id: string;
  title: string;
  url: string;
  shared_by: string;
  description: string;
}

export interface MovieContextData {
  movies: Movie[];
  updateMovies: (movies: Movie[]) => void;
}
