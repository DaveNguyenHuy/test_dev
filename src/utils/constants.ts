export const ROUTES = {
  home: '/',
  share: '/share'
};

export const YouTubeGetID = (url: string) => {
  let ID: string | string[] = '';
  const parsedUrl = url.replace(/(>|<)/gi, '').split(/(vi\/|v=|\/v\/|youtu\.be\/|\/embed\/)/);
  if (parsedUrl[2] !== undefined) {
    return parsedUrl[2].split(/[^0-9a-z_-]/i)[0];
  }
  return ID;
};
