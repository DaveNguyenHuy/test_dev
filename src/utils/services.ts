import { collection, getDocs, query } from '@firebase/firestore';
import { db } from './firebase';

const moviesCollection = collection(db, 'movies');

export const fetchMovies = async () => {
  try {
    const q = query(moviesCollection);
    const querySnapshot = await getDocs(q);
    const data: any = [];
    querySnapshot.forEach((doc) => {
      data.push({
        id: doc.id,
        ...doc.data()
      });
    });
    return data;
  } catch (error) {
    return [];
  }
};
