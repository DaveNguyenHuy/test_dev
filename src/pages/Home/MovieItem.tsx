import { FC, useContext, useState } from 'react';
import { BiLike, BiDislike } from 'react-icons/bi';
import { MoviesContext } from '../../App';
import { updateMovieService } from '../../utils/services';
import { Movie, Like } from '../../utils/types';

interface Props {
  movie: Movie;
  updateMovie: (movie: Movie) => void;
}

export const MovieItem: FC<Props> = ({ movie, updateMovie }) => {
  const { currentUser } = useContext(MoviesContext);
  const [loading, setLoading] = useState(false);

  const like: Like | undefined = (movie.likes || []).find((it) => it.email === currentUser?.email);
  const { isLike, isDislike } = { isLike: !!like && like.like, isDislike: !!like && like.dislike };

  const toggleLike = async (key: 'like' | 'dislike') => {
    if (loading) return;
    const newMovie = { ...movie };
    if (!like) {
      const newLike = {
        email: currentUser.email,
        [key]: true
      };
      newMovie.likes = [...(newMovie.likes || []), newLike];
    } else {
      const other = key === 'like' ? 'dislike' : 'like';
      like[key] = !like[key];
      if (like[other]) like[other] = false;
    }
    setLoading(false);
    await updateMovieService(newMovie);
    updateMovie(newMovie);
    setLoading(false);
  };

  return (
    <div className="flex space-x-8 max-w-screen-lg w-full">
      <iframe
        src={movie.url}
        title={movie.title}
        height="240"
        width="360"
        className="border-gray-800 border flex-shrink-0 rounded-sm"
      />
      <div className="flex-1">
        <div className="flex justify-between w-full">
          <div className="w-4/5">
            <h2 data-testid="videoTitle" className="text-2xl font-bold">
              {movie.title}
            </h2>
            <p data-testid="videoSharedBy" className="text-gray-500">
              Shared by: {movie.shared_by}
            </p>
          </div>
          {currentUser && (
            <div className="space-x-2">
              <span
                className="p-1.5 cursor-pointer h-auto hover:bg-gray-200 rounded-full inline-block"
                onClick={() => toggleLike('like')}
              >
                <BiLike color={isLike ? 'blue' : 'black'} className="w-6 h-6" />
              </span>
              <span
                className="p-1.5 cursor-pointer h-auto hover:bg-gray-200 rounded-full inline-block"
                onClick={() => toggleLike('dislike')}
              >
                <BiDislike color={isDislike ? 'blue' : 'black'} className="w-6 h-6" />
              </span>
            </div>
          )}
        </div>
        <p className="mt-2 text-lg">Description:</p>
        <p data-testid="videoDescription" className="movie-description">
          {movie.description}
        </p>
      </div>
    </div>
  );
};
