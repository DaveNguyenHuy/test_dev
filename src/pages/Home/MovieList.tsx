import { MovieItem } from './MovieItem';

export const MovieList = () => {
  return (
    <div className="flex flex-col items-center pt-8 space-y-8">
      <MovieItem />
      <MovieItem />
      <MovieItem />
      <MovieItem />
      <MovieItem />
      <MovieItem />
      <MovieItem />
      <MovieItem />
    </div>
  );
};
