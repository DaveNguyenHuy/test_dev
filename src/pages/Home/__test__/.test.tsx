import { render, cleanup, screen } from '@testing-library/react';
import { within } from '@testing-library/dom';
import { Movie } from '../../../utils/types';
import { MovieItem } from '../MovieItem';
import { MovieList } from '../MovieList';

beforeEach(() => {
  cleanup();
});

describe('testing home components', () => {
  test('Movie item should render content', () => {
    const data: Movie = {
      id: '1234',
      title: 'testing video',
      url: 'https://www.youtube.com/embed/test_id',
      description: 'testing video',
      shared_by: 'test@gmail.com'
    };

    const { getByTestId, getByTitle, container } = render(<MovieItem movie={data} />);

    expect(container.firstChild).toMatchSnapshot();

    expect(getByTitle(data.title)).toBeInTheDocument();
    const { getByText } = within(getByTestId('videoTitle'));
    expect(getByText(/testing video/i)).toBeInTheDocument();
    expect(getByTestId('videoSharedBy')).toHaveTextContent(data.shared_by);
    expect(getByTestId('videoDescription')).toHaveTextContent(data.description);
  });

  test('Movie list should render empty text', () => {
    const { getByText } = render(<MovieList movies={[]} />);
    expect(getByText('There is no movies...')).toBeInTheDocument();
  });

  test('Movie list should render content', () => {
    const movies: Movie[] = [
      {
        id: '1',
        title: 'movie 01',
        description: 'movie 01 description',
        shared_by: 'test@example.com',
        url: 'https://www.youtube.com/embed/test_1'
      },
      {
        id: '2',
        title: 'movie 02',
        description: 'movie 02 description',
        shared_by: 'test@example.com',
        url: 'https://www.youtube.com/embed/test_2'
      }
    ];

    const { container } = render(<MovieList movies={movies} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
