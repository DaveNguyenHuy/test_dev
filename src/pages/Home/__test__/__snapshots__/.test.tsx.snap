// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing home components Movie item should render content 1`] = `
<div
  class="flex space-x-8 max-w-screen-lg w-full"
>
  <iframe
    class="border-gray-800 border flex-shrink-0 rounded-sm"
    height="240"
    src="https://www.youtube.com/embed/test_id"
    title="testing video"
    width="360"
  />
  <div
    class="flex-1"
  >
    <div
      class="flex justify-between w-full"
    >
      <div
        class="w-4/5"
      >
        <h2
          class="text-2xl font-bold"
          data-testid="videoTitle"
        >
          testing video
        </h2>
        <p
          class="text-gray-500"
          data-testid="videoSharedBy"
        >
          Shared by: 
          test@gmail.com
        </p>
      </div>
    </div>
    <p
      class="mt-2 text-lg"
    >
      Description:
    </p>
    <p
      class="movie-description"
      data-testid="videoDescription"
    >
      testing video
    </p>
  </div>
</div>
`;

exports[`testing home components Movie list should render content 1`] = `
<div
  class="flex flex-col items-center pt-8 space-y-8"
>
  <div
    class="flex space-x-8 max-w-screen-lg w-full"
  >
    <iframe
      class="border-gray-800 border flex-shrink-0 rounded-sm"
      height="240"
      src="https://www.youtube.com/embed/test_1"
      title="movie 01"
      width="360"
    />
    <div
      class="flex-1"
    >
      <div
        class="flex justify-between w-full"
      >
        <div
          class="w-4/5"
        >
          <h2
            class="text-2xl font-bold"
            data-testid="videoTitle"
          >
            movie 01
          </h2>
          <p
            class="text-gray-500"
            data-testid="videoSharedBy"
          >
            Shared by: 
            test@example.com
          </p>
        </div>
      </div>
      <p
        class="mt-2 text-lg"
      >
        Description:
      </p>
      <p
        class="movie-description"
        data-testid="videoDescription"
      >
        movie 01 description
      </p>
    </div>
  </div>
  <div
    class="flex space-x-8 max-w-screen-lg w-full"
  >
    <iframe
      class="border-gray-800 border flex-shrink-0 rounded-sm"
      height="240"
      src="https://www.youtube.com/embed/test_2"
      title="movie 02"
      width="360"
    />
    <div
      class="flex-1"
    >
      <div
        class="flex justify-between w-full"
      >
        <div
          class="w-4/5"
        >
          <h2
            class="text-2xl font-bold"
            data-testid="videoTitle"
          >
            movie 02
          </h2>
          <p
            class="text-gray-500"
            data-testid="videoSharedBy"
          >
            Shared by: 
            test@example.com
          </p>
        </div>
      </div>
      <p
        class="mt-2 text-lg"
      >
        Description:
      </p>
      <p
        class="movie-description"
        data-testid="videoDescription"
      >
        movie 02 description
      </p>
    </div>
  </div>
</div>
`;
